docs/rfcs/1011-non-blocking-invite-transport.md

```diff
@@ -0,0 +1,915 @@
+
+Note that this set may evolve overtime as users can be revoked and user profiles can change.
+
+
+## Goals and Non-Goals
+
+To sum it up, the goal is to design a new invite transport that:
+- allows for network failures by removing the simultaneity constraints of Parsec v2 and using non-blocking HTTP requests
+- adds new semantics to manage attempts distinctly and allow for cancelling them with a reason that's communicated to the other peer
+- decouples the concept of greeter from the invitation author so that a user can be greeted by any administrator
+
+
+## Terminology
+
+Both terms `invite` and `invitation` are used in the code base. There are basically three usage of those words:
+
+1. `invite` as verb, as in "an administrator invites a user to join an organization"
+  * example: the text for `NewUserInvitationError::NotAllowed` is `Not allowed to invite a user`
+
+2. `invite` as a noun, referring to the invite system in general
+  * example: the commands related to the invite system are named `invite_*`
+
+3. `invitation`, referring to a specific invitation that has to managed
+  * example: most invite commands have return status called `invitation_not_found` or `invitation_deleted`
+
+In order to avoid too many changes in the data model, we'll stick to this terminology and apply it in the implementation of this RFC.
+
+## Semantics
+
+This section offers a new semantics proposal, without going into the technical aspects of the exposed routes and their implementations.
+
+Each concept is described with their identifier, immutable properties, internal state, corresponding getters and exposed methods.
+
+
+### Invitation
+
+An invitation is used to verify the identity of a user (called the claimer) on a specific device. The users allowed to verify the identity are called greeters. The invitation can be of 3 types:
+- User invitation
+- Device invitation
+- Shared recovery (shamir) invitation
+
+An invitation is identified by:
+- organization id
+- invitation token
+
+Immutable properties:
+- type (user, device, shamir)
+- created-on timestamp
+- created-by author
+- claimer id:
+  * email address (for user invitation)
+  * or user id (for device or shamir invitation)
+
+Mutable state:
+- state: `PENDING` | `CANCELLED` | `COMPLETED`
+
+Getters:
+- get greeters
+  * for user invitation: all administrators
+  * for device invitation: same as claimer id
+  * for shamir invitation: all recipients
+- get channel for greeter
+
+Methods:
+- create invitation
+- cancel invitation
+- complete invitation
+
+
+### Channel
+
+A channel is dedicated to the verification of the claimer identity by a specific greeter.
+
+A channel is identified by:
+- organization id
+- invitation token
+- greeter id
+
+Mutable properties:
+- attempts ids: list of attempts
+   * Invariant: most recent attempt is active, all the rest is cancelled
+
+Getters:
+- get active attempt
+
+Methods:
+- cancel active attempt (and create a new one)
```

Comment:
Is the creation of a new attempt optional ?

Reply:
My idea was to have the following invariant:
- there always exactly one active attempt by channel and it is the most recent
In order to respect the invariant, cancelling the active attempt implies the creation of a new one, atomically.
In practice, the creation of the active attempt could also be done lazily when needed.

But that's a proposal, there might be another way to handle this.

Reply:
With the current proposal, created a new active attempts could not be done lazily.
As you said it would break your Invariant that always the most recent is activate.

But I think that would be doable if we had 2 fields:

```typescript
interface Attempts {
	activate_attempt: ID | Null;
	cancelled_attempts: ID[];
}
```

Reply:
> With the current proposal, created a new active attempts could not be done lazily.

How so? IMO, as long as methods like `get_active_attempt` always return an active attempt, the invariant is respected. That means that `get_active_attempt` should be able to create the active attempt if it does not exist. That's what I meant by lazy creation. But this is an implementation detail, and we could also choose to atomically create a new attempt when the previous one is cancelled. Whether the first solution or the second solution is implemented it does not matter, it should be transparent for the caller that the invariant is respected in both cases.

> But I think that would be doable if we had 2 fields:

I think I shouldn't have put `attempts ids` as a mutable property of the `channel` because in practice, we'll have a postgre table called `attempts` for storing all attempts id and their internal state. That means that methods like `get_active_attempt` or `cancel_active_attempt` are merely requests over `attempts` table. The invariant is an invariant to maintain in the `attempts` table, not in the internal state of the `channel`.

So all things considered, a `channel` have no internal state meaning it doesn't have a corresponding table either. However, I think it's still useful that the concept exists to properly name things in the code.